Semaphore: remove 16-bit limit

kjbracey opened the pull request:
### Description

Semaphore had a historical 16-bit maximum count limit - this was odd, as the initial count was 32-bit.

Remove the limitation - change constructors to take uint32_t, and make the default limit 0xffffffff rather than 0xffff.

Limits of overload resolution prevent adding new constructors while retaining the old ones.

Addresses issue raised in #10216.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [X] Breaking change

### Release Notes

* `Semaphore` now permits the full 32-bit maximum count available in CMSIS RTOS 2. The change to its available constructors breaks binary compatibility, but there should be no source compatibility issues.
* The default maximum count for a `Semaphore` is now 0xffffffff rather than 0xffff.

kjbracey committed:
Semaphore: remove 16-bit limit

Semaphore had a historical 16-bit maximum count limit - this was odd, as
the initial count was 32-bit.

Remove the limitation - change constructors to take uint32_t, and make
the default limit 0xffffffff rather than 0xffff.

Limits of overload resolution prevent adding new constructors while
retaining the old ones.

Addresses issue raised in #10216.

desowin commented on rtos/Semaphore.h:
```diff
@@ -49,15 +49,15 @@ class Semaphore : private mbed::NonCopyable<Semaphore> {
       @note You cannot call this function from ISR context.
     */
-    Semaphore(int32_t count = 0);
+    Semaphore(uint32_t count = 0);
```
I believe the reason why count was `int32_t` is due to the fact that `wait()` returns the number of available tokens or -1 on invalid parameters. With count being `uint32_t` you cannot differentiate between the invalid parameters and (very unlikely) case where you have (2^32)-1 tokens available (on U2 complement systems).

kjbracey replied:
Well spotted. That particular `wait` API is pretty horrible, and I'm intending to deprecate it anyway in favour of a more rational set of `try_acquire`, `acquire`, `acquire_for` and `acquire_until` (aligning with the current state of `Mutex::lock`).

Maybe we could fudge `wait` so it just clamps to `0x7FFFFFFF`? You can't really distinguish different positive return values usefully in a non-racey way anyway.

desowin replied:
I think clamping `wait()` output to `0x7FFFFFFF` is really nasty. I'd just go with `int32_t` for the count and maximum count instead.

kjbracey replied:
What's nasty is returning a count at all - you can't really distinguish between any return values in the range 1 to 0x7FFFFFFF, as the instantaneous read is not atomic or synchronised in any way. All you know for certain is that you acquired 1.

`wait` and its count returning is on the chopping block for deprecation - `acquire` will not return a count. So I'm minded to ignore the current flaws of `wait`, rather than work around them.

kjbracey committed:
Limit Semaphore::wait() return so it is positive

If a Semaphore has a huge count, it could overflow the int32_t return of
Semaphore::wait().
